pkg/abstractions/machine/machine.go

```diff
@@ -0,0 +1,41 @@
+package machine
+
+import (
+	"context"
+
+	"github.com/beam-cloud/beta9/pkg/repository"
+	"github.com/beam-cloud/beta9/pkg/types"
+	pb "github.com/beam-cloud/beta9/proto"
+)
+
+type MachineService interface {
+	pb.MachineServiceServer
+	GetGPUAvailability(ctx context.Context, req *pb.GetGPUAvailabilityRequest) (*pb.GetGPUAvailabilityResponse, error)
+}
+
+type MachineRedisService struct {
+	pb.UnimplementedMachineServiceServer
+	config       map[string]types.WorkerPoolConfig
+	providerRepo repository.ProviderRepository
+}
+
+func NewMachineRedisService(config map[string]types.WorkerPoolConfig, providerRepo repository.ProviderRepository) MachineService {
+	return &MachineRedisService{
+		providerRepo: providerRepo,
+	}
+}
+
+func (m *MachineRedisService) GetGPUAvailability(ctx context.Context, in *pb.GetGPUAvailabilityRequest) (*pb.GetGPUAvailabilityResponse, error) {
```

Comment:
I'm a little unsure if this is an "abstraction" / belongs here. I guess the rationale is that since it's exposed in the SDK, its an abstraction -- not sure though.

Reply:
Feels like a core api in a sense, but we really don't have that concept - at least not categorized. I also think if we define a new core api type of category, then we'll need to provide a stronger definition of what an abstraction is vs whatever other category we introduce.

But if we say an abstraction is something exposed to the SDK - like you said - then I think it makes sense to keep machines as an abstraction.